### ISSUE TITLE
impress.js status update

### ISSUE THREAD

**Issue by henrikingo** [impress/impress.js#640]
Hi all

I wanted to give all watchers a bit of a heads up before the big merge hits the master.

As many of you know, over the past 2 years I developed a plugin based architecture as well as lots of new features for impress.js. For various reasons these were not merged back upstream, but I'm glad to tell you that this is finally happening. All of my work is now fully merged into and available as a preview in [the dev branch](https://github.com/impress/impress.js/tree/dev) 

**Unless there are some valid objections, then one week from now, on Sunday Nov 12, I will merge the dev branch into master** and tag it as 1.0.0-beta1.

That will also cause most of the open issues and pull requests to be closed after which I will also open several new ones (such as from the list in #626). From now on I'll also start marking suitable issues with the github recommended labels "help wanted" and "good first issue" so that if there is anyone who would like to contribute to a popular JavaScript project, that will now be possible.

**Comment by FagnerMartinsBrack** [comment]
Can you create a PR for that merge and leave it open? One week may be too soon, can we make it at least one month?

**Comment by henrikingo** [comment]
> Can you create a PR for that merge and leave it open so that the diff can be clearly exposed?

No! Surely you're joking? The diff is the aggregate of everything you reviewed in the past PRs, 2 years worth of code. If you want to review, go back to the PRs I've sent in the past weeks. All merge-* branches are still open.

> One week may be too soon given the number of changes and the lack of test coverage, can we make it at least one month?

There's no rush. Of course, that would imply nobody else should merge anything to master either, but given that this isn't happening anyway, not a huge problem.

That said, what are we worried about exactly? Stable releases are tagged, so if users want stable, they should use that. If they use master, they probably want the latest and greatest, so why not finally give them that? Note that the new stuff in the dev branch is not new or untested code any more than master: it's been in my fork for 2 years and has lots of users already.

And conversely, if we wait longer, what do we gain? Do you intend to actively test the dev branch? Does someone else? I can wait, but what am I waiting for?

**Comment by FagnerMartinsBrack** [comment]
> If you want to review, go back to the PRs I've sent in the past weeks. All merge-* branches are still open.

As [mentioned before](https://github.com/impress/impress.js/pull/625#issuecomment-334722766), the idea was to review the whole dev branch once it finished being copied over.

Those PRs were not created with an [atomic concern](https://medium.com/@fagnerbrack/one-pull-request-one-concern-e84a27dfe9f1), that makes it extremely hard to review because it trades off quality of review for speed of delivery and keeping the context of the code in the author's head. Creating structured PRs allow for somebody else in the future (maybe another contributor) to understand the reasoning behind each feature. For example, if adding a plugin, it's possible to add only the API in impress.js that is necessary for that plugin to work. That will ensure the API doesn't expose anything other than what's necessary to satisfy that use case, reviewing that we efficiently avoid clutter and [remove code](https://medium.freecodecamp.org/code-that-dont-exist-is-the-code-you-don-t-need-to-debug-88985ed9604).

I've mentioned this before, but it seems that [@bartaz and yourself are ok with the tradeoff of quality of review (and the efforts to do so) for speed of delivery](https://github.com/impress/impress.js/pull/625#discussion_r143154698). I would argue that's not something that's worth it because it will make the codebase rigid, but I won't insist on it if we're all conscious of this tradeoff and that's the only way forward.

> ... it's been in my fork for 2 years and has lots of users already

Does that mean you believe the code is very good? If that's so, why create a Pull Request? What needs to be reviewed exactly? Shouldn't you just commit to master?

That statement [can be a fallacy](https://en.wikipedia.org/wiki/Argument_from_authority).

> And conversely, if we wait longer, what do we gain? Do you intend to actively test the dev branch? Does someone else? I can wait, but what am I waiting for?

I intended to look at the `dev` branch against master and review it again, [as mentioned a month ago](https://github.com/impress/impress.js/pull/625#issuecomment-334722766). It also opens the possibility for a wider audience to peek in. I'm not sure how active are the watchers of this repo, but if there's a slight chance to have [more eyeballs](https://en.wikipedia.org/wiki/Linus%27s_Law) into this I would argue we should do it.

Yes, I could just look at the [`master...dev` diff](https://github.com/impress/impress.js/compare/master...dev). However, a Pull Request creates noise and maybe (just maybe) somebody else can jump in and add something useful to the discussion. Would that possibility be worth it?

> Note that the new stuff in the dev branch is not new or untested code any more than master: it's been in my fork for 2 years and has lots of users already.

From what I've seen so far, even though the merge promises to close a lot of feature requests from this repo, it adds a lot of code that will now be exposed to a much bigger audience. The code might be stable and working today, however, it seems that it provides **stability** and **more features** at the cost of:

* **Rigidness**. Due to the increase of software entropy and the lack of tests and [self-documentation](https://medium.com/@fagnerbrack/code-comment-is-a-smell-4e8d78b0415b), the code becomes more rigid. That means that new contributors will be afraid to change the code more than before. Because it didn't have tests before it actually makes this problem worse, not better.
* **Less developer contribution**. Future contributors won't be able to easily create Pull Requests for new features or fixes to be added to the project and it might stale or only be contributed by @henrikingo who has the context of how it works. Right now, the merge increases @henrikingo's [Bus Factor](https://en.wikipedia.org/wiki/Bus_factor) in the project. Tests can also serve as a documentation of how the impress API works in code (the so-called "first client" of the API), if there are no tests for contributors to look at, new contributors will have to spend so much time to understand everything that they will most likely give up.

My understanding at this moment (and I partly blame on how the format of the review was done) is that this merge optimizes feature delivery by sacrificing contribution and the flexibility of impress.js core.

**Of course, I can be wrong...**

I don't have the whole context of the features added. That's why I intended to look at the `dev` branch as a whole in a Pull Request and see if that would make things clearer.

I was only able to look at small isolated and non-cohesive snapshots of changes.

**But if I'm right...**

If I'm right and @bartaz and @henrikingo are ok with the tradeoffs I listed there, then I won't insist on it and it should be good to go to `master`.

I just want to make sure we make the conscious decision because I also want impress.js to succeed, not to become worse from a maintainability perspective.

I believe this is a reasonable observation, what do you think?

**Comment by henrikingo** [comment]
> As mentioned before, the idea was to review the whole dev branch once it finished being copied over.

You must have misunderstood. Those pull requests were pull requests created specifically for the purpose of review. You participated in reviewing them. Now review is over.

> Those PRs were not created with an atomic concern,

"atomic concern"? Those PRs were created in the format requested by Bartek, and I spent considerable amount of work to do that so that he could review them. If you want to review smaller units of work, you're free to review the almost 200 commits in my fork that are the true history of this work. In fact, this request for prolonged review is particularly outrageous coming from you, as you have been well aware of my fork, and as maintainer it would have been your job to work on getting those features merged upstream, but you did not do that.

Luckily, now I'm active maintainer of impress.js, and progress will not be blocked anymore. I have now merged the dev branch, which means 1.0.0-beta1 is released, and also this unblocks master branch so I can now focus on another PR already in the queue, rather than having it blocked by this discussion for a month too. This will be modus operandi going forward.

**Comment by FagnerMartinsBrack** [comment]
> In fact, this request for prolonged review is particularly outrageous coming from you, as you have been well aware of my fork, and as maintainer it would have been your job to work on getting those features merged upstream, but you did not do that.

There's no need for ad hominem, it removes the focus of the discussion. You have never responded to my comment, you just attacked me for no reason.

I've been aware that you created a fork because you didn't like my reviews, but I never had time to look at it. When you were contributing to this repo, you've never implemented a single suggestion I've offered, you just ignored it, forked off and started adding comments to the issues of this repo referring to your fork.

It's nobody's job to merge somebody else's fork. I'm not paid to do Open Source and I'm not paid to listen to this. I already contributed a hell lot of time to Open Source and I have no company to back me up, I do it in my own time. Even though I couldn't do much, I still managed to create [The Reference API](https://github.com/impress/impress.js/blob/master/DOCUMENTATION.md) for impress.js because it didn't have any and triage almost 300 issues and 200 Pull Requests down into what you can see today.

It reminds me of the post [The Open Source Community Has No Place for Disrespect](https://medium.com/@d4nyll/the-open-source-community-have-no-place-for-disrespect-70c85d473332). The Open Source community also has no right to demand anything from somebody who does the job for free. This behavior is poisoning and your tone is also.

My comment still stands and is just a suggestion, not a requirement.

Let's have a reasonable, mature conversation and stop the ad-hominem.

**Comment by henrikingo** [comment]
> When you were contributing to this repo, you've never implemented a single suggestion I've offered

That is not true. Eventually however I stopped implementing them, because it wasn't leading to code being pushed to the repo. Of course, it is also true I disagree with most of your suggestions both now and then. 

So what has happened instead is that Bartek has come back to review and approve my PRs, and thanks to this now my work is finally merged to the upstream repo. Also thanks to this, I'm now active maintainer, so going forward I will be processing pull requests more expediently and hopefully repo will be much more active!

The code that was discussed in this issue is merged, so please don't reopen it anymore. You can continue to comment without reopening. But I will be focusing on working on code, and I recommend the same to you.

**Comment by bartaz** [comment]
@henrikingo @FagnerMartinsBrack Sorry that I missed this thread earlier, last week was quite busy for me.

Let's try not to mix too much emotions in this, as it's not very helpful.
There may be some misunderstandings between us, disagreements, but let's try not take it too personally because it's unnecessary for anyone. Every single one of us contributes their free time to this in their own way, and let's respect that.

For the sake of this thread about merging dev branch to master.
I was hoping (I believe similar to @FagnerMartinsBrack) that we discuss and address at least some of things mentioned in #626 before the merge. But as I understand @henrikingo intention is to address them now, so I assume that it will happen before moving out of 'beta'.

And because of master being in 'beta' I think it would be nice to have a note in readme that would point to last 'stable' release.

About the comments about rigidness and outside contributions from @FagnerMartinsBrack.
You are probably right that this may happen. But it will be hard to avoid when codebase gets bigger and gets more features. Naturally it will be more 'scary' to newcomers.
Now with plugin architecture maybe it will change once they settle in and are documented enough. But to see that we probably need to wait couple of weeks or months for some contributions to appear. Personally I don't expect much, because even that impress.js seemed to be quite popular as a project there was not really a 'community' around that. First I was the 'bus factor' and project stopped when I stopped. Having @henrikingo as a contributor is at least some way forward, because of his willingness to contribute code and maintain it.
It's better to have one developer active then endlessly discussing how to make code more approachable to contributors and have no contributors at all (in my opinion).

And, getting back to the emotions from the beginning - if we want to attract more people to the project and have more contributions we need to try to be friendly and respectful for them. So let's start with trying this between ourselves :)

So to summarise, @henrikingo - what are the next steps?

I'd like to see before we remove beta flag:

- issues/PRs addressing items from #626.
- In the meantime possibly going through the backlog and looking what is fixed already by plugins and what can now be addressed.
- looking into README and other docs to see if they are up to date with recent changes

Anything else?

**Comment by FagnerMartinsBrack** [comment]
> But it will be hard to avoid when codebase gets bigger and gets more features.

It's very easy to avoid rigidness if we apply common software development practices, I can help with some input on that for future PRs. My point is just that we need to be conscious of the tradeoffs if we don't do it right now so that we can make better decisions and improve the codebase in the future.

-----

As from this point, I don't want to read emotional ad-hominem anymore, I've read enough. Leave emotions at home and start [arguing against the content, not the person](https://medium.com/@fagnerbrack/personal-experience-doesnt-matter-cb060b42f99a), just how @bartaz did with his comment.

**Comment by henrikingo** [comment]
Just a short note to anyone who might read this and wonder what happened. Above thread caused me to write Bartek in private. There was another email between Bartek and Fagner, that I was not a party to.

Between all the distraction, Bartek's comment also included some valid proposals and questions. However, we agree it is not worth continuing this thread. So rather than answering to those here, future steps will become evident in their respective issue threads and commits.
